patroni/dcs/etcd3.py

```diff
@@ -302,8 +306,11 @@ def authenticate(self) -> bool:
             self._token = response.get('token')
         return old_token != self._token
 
+    def create_auth_retry(self) -> Retry:
+        return Retry(max_tries=-1, retry_exceptions=(RevisionOfAuthStoreIsOld,))
```

Comment:
Thanks for detailed description of the issue! 
Unfortunately we can't afford infinite retries, and there are chances of it to happen if someone will modify users with the high rate.

In the `kwargs` of the `handle_auth_errors()` method there could be a `retry` object, that should be used:
1. If there is no `retry` argument we can just give up immediately and don't perform reauthentication at all. The retry will happen naturally, on the next hear beat cycle.
2. If there is a `retry` object, it should be used to make sure that the total execution time doesn't exceed the `Retry.deadline` (e.g. `retry_timeout`).

Also, as of now the successful `authenticate()` method restarts a watcher thread, but in case of `RevisionOfAuthStoreIsOld` we don't have to do it, because it continues to work normally.

I'll try to come up with the proper fix and retry policies today.

Reply:
Something like this: https://github.com/zalando/patroni/pull/2913/commits/47628a7f1d8a35e51a147911a6bc7a8ccdc84af0

Reply:
Yea, that makes sense. Thank you for making this fix! I'll close this PR since https://github.com/zalando/patroni/pull/2913 exists then.